File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Recipient/RecipientHandler.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.Collections.Generic;
using System.Text.Json;
using HealthFlow.Hfcx.Sdk.Auth;
using HealthFlow.Hfcx.Sdk.Exceptions;
using HealthFlow.Hfcx.Sdk.Logging;
using HealthFlow.Hfcx.Sdk.Protocol;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace HealthFlow.Hfcx.Sdk.Recipient;

/// <summary>
/// Orchestrates the four-layer HFCX recipient pipeline:
/// <c>Bearer → Headers → Fhir → Egyptian</c>. Each layer is
/// independently toggleable.
/// </summary>
/// <remarks>
/// Sister to Java's <c>RecipientHandler</c> and Python's
/// <c>RecipientHandler</c>. The SDK does NOT ship a default
/// trust-everything <see cref="IBearerTokenValidator"/> — enabling
/// <see cref="Layer.Bearer"/> without configuring one fails fast at
/// construction.
/// </remarks>
public sealed class RecipientHandler
{
    private readonly IReadOnlySet<Layer> _enabledLayers;
    private readonly InboundDecryptor _decryptor;
    private readonly IBearerTokenValidator? _bearerValidator;
    private readonly HeaderValidator? _headerValidator;
    private readonly FhirValidator? _fhirValidator;
    private readonly EgyptianBundleValidator? _egyptianValidator;
    private readonly ILogger _log;

    /// <summary>Construct the handler. Pass <paramref name="enabledLayers"/> as null for all four.</summary>
    public RecipientHandler(
        ILocalKeyProvider keyProvider,
        string? localParticipantCode = null,
        IBearerTokenValidator? bearerTokenValidator = null,
        IReadOnlySet<Layer>? enabledLayers = null,
        Func<DateTimeOffset>? clock = null,
        TimeSpan? timestampTolerance = null,
        ILogger<RecipientHandler>? logger = null)
    {
        ArgumentNullException.ThrowIfNull(keyProvider);

        _enabledLayers = enabledLayers
            ?? new HashSet<Layer> { Layer.Bearer, Layer.Headers, Layer.Fhir, Layer.Egyptian };

        if (_enabledLayers.Contains(Layer.Headers) && string.IsNullOrEmpty(localParticipantCode))
        {
            throw new ArgumentException(
                "localParticipantCode is required when Layer.Headers is enabled",
                nameof(localParticipantCode));
        }

        if (_enabledLayers.Contains(Layer.Bearer) && bearerTokenValidator is null)
        {
            throw new ArgumentException(
                "Layer.Bearer is enabled but no IBearerTokenValidator was supplied. "
                + "The SDK does NOT ship a default trust-everything validator; "
                + "configure one or disable Layer.Bearer explicitly.",
                nameof(bearerTokenValidator));
        }

        _decryptor = new InboundDecryptor(keyProvider);
        _bearerValidator = bearerTokenValidator;
        _headerValidator = _enabledLayers.Contains(Layer.Headers)
            ? new HeaderValidator(localParticipantCode!, clock, timestampTolerance)
            : null;
        _fhirValidator = _enabledLayers.Contains(Layer.Fhir) ? new FhirValidator() : null;
        _egyptianValidator = _enabledLayers.Contains(Layer.Egyptian) ? new EgyptianBundleValidator() : null;
        _log = logger ?? NullLogger<RecipientHandler>.Instance;
    }

    /// <summary>The set of layers currently enabled.</summary>
    public IReadOnlySet<Layer> EnabledLayers => _enabledLayers;

    /// <summary>
    /// Run the pipeline against an inbound request. Throws the
    /// appropriate typed <see cref="HfcxException"/> on validation
    /// failure; returns <see cref="RecipientResult"/> on success.
    /// </summary>
    public RecipientResult Handle(
        string? authorizationHeader,
        IReadOnlyDictionary<string, string> protocolHeaders,
        string requestBody)
    {
        ArgumentNullException.ThrowIfNull(protocolHeaders);
        ArgumentNullException.ThrowIfNull(requestBody);

        var correlationId = protocolHeaders.TryGetValue(ProtocolHeaders.CorrelationId, out var cid)
            ? cid
            : "no-correlation-id";

        using (CorrelationId.Scope(correlationId))
        {
            _log.LogInformation(
                "recipient: handling inbound request with layers {Layers}",
                string.Join(",", _enabledLayers));

            if (_enabledLayers.Contains(Layer.Bearer))
            {
                _bearerValidator!.Validate(authorizationHeader);
            }

            if (_enabledLayers.Contains(Layer.Headers))
            {
                _headerValidator!.Validate(protocolHeaders);
            }

            var jwe = ExtractPayload(requestBody);
            var decrypted = _decryptor.Decrypt(jwe);

            _fhirValidator?.Validate(decrypted);
            _egyptianValidator?.Validate(decrypted);

            _log.LogInformation("recipient: accepted (all enabled layers passed)");
            return new RecipientResult(decrypted, protocolHeaders, correlationId);
        }
    }

    private static string ExtractPayload(string requestBody)
    {
        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(requestBody);
        }
        catch (JsonException ex)
        {
            throw new EnvelopeMalformedJsonException(
                $"Request body is not valid JSON: {ex.Message}");
        }

        using (doc)
        {
            if (doc.RootElement.ValueKind != JsonValueKind.Object)
            {
                throw new EnvelopeMissingPayloadException(
                    "Request body envelope is not a JSON object");
            }

            if (!doc.RootElement.TryGetProperty("payload", out var payload)
                || payload.ValueKind != JsonValueKind.String
                || string.IsNullOrEmpty(payload.GetString()))
            {
                throw new EnvelopeMissingPayloadException(
                    "Request body envelope missing required 'payload' field");
            }

            return payload.GetString()!;
        }
    }
}

File: sdk-javascript/src/recipient/RecipientHandler.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import type { BearerTokenValidator } from '../auth/BearerTokenValidator.js';
import {
  EnvelopeMalformedJsonError,
  EnvelopeMissingPayloadError,
} from '../exceptions/HfcxError.js';
import { runWithCorrelationId } from '../logging/CorrelationId.js';
import { PROTOCOL_HEADER_CORRELATION_ID } from '../protocol/ProtocolHeaders.js';
import { EgyptianBundleValidator } from './EgyptianBundleValidator.js';
import { FhirValidator } from './FhirValidator.js';
import { HeaderValidator } from './HeaderValidator.js';
import { InboundDecryptor } from './InboundDecryptor.js';
import { ALL_LAYERS, type Layer } from './Layer.js';
import type { LocalKeyProvider } from './LocalKeyProvider.js';
import type { RecipientResult } from './RecipientResult.js';

export interface RecipientHandlerOptions {
  keyProvider: LocalKeyProvider;
  /** Required when {@link Layer.HEADERS} is enabled. */
  localParticipantCode?: string;
  /** Required when {@link Layer.BEARER} is enabled. */
  bearerTokenValidator?: BearerTokenValidator;
  /** Defaults to all four layers. Pass an empty set to disable everything. */
  enabledLayers?: ReadonlySet<Layer>;
  /** Test seam: wall-clock for the timestamp tolerance check. */
  clock?: () => Date;
  /** Default ±5 min. */
  timestampToleranceMs?: number;
}

/**
 * Orchestrates the four-layer HFCX recipient pipeline:
 * `BEARER → HEADERS → FHIR → EGYPTIAN`. Each layer is independently
 * toggleable.
 *
 * Sister to Java's `RecipientHandler`, Python's `RecipientHandler`,
 * and .NET's `RecipientHandler`. The SDK does NOT ship a default
 * trust-everything {@link BearerTokenValidator} — enabling
 * `Layer.BEARER` without configuring one fails fast at construction.
 */
export class RecipientHandler {
  readonly enabledLayers: ReadonlySet<Layer>;
  private readonly decryptor: InboundDecryptor;
  private readonly bearerValidator: BearerTokenValidator | undefined;
  private readonly headerValidator: HeaderValidator | undefined;
  private readonly fhirValidator: FhirValidator | undefined;
  private readonly egyptianValidator: EgyptianBundleValidator | undefined;

  constructor(options: RecipientHandlerOptions) {
    if (!options.keyProvider) throw new TypeError('keyProvider is required');

    const layers = options.enabledLayers ?? new Set<Layer>(ALL_LAYERS);
    if (layers.has('HEADERS') && !options.localParticipantCode) {
      throw new TypeError('localParticipantCode is required when Layer.HEADERS is enabled');
    }
    if (layers.has('BEARER') && !options.bearerTokenValidator) {
      throw new TypeError(
        'Layer.BEARER is enabled but no BearerTokenValidator was supplied. ' +
          'The SDK does NOT ship a default trust-everything validator; ' +
          'configure one or disable Layer.BEARER explicitly.',
      );
    }

    this.enabledLayers = layers;
    this.decryptor = new InboundDecryptor(options.keyProvider);
    this.bearerValidator = options.bearerTokenValidator;
    this.headerValidator = layers.has('HEADERS')
      ? new HeaderValidator(options.localParticipantCode!, {
          ...(options.clock !== undefined ? { clock: options.clock } : {}),
          ...(options.timestampToleranceMs !== undefined
            ? { timestampToleranceMs: options.timestampToleranceMs }
            : {}),
        })
      : undefined;
    this.fhirValidator = layers.has('FHIR') ? new FhirValidator() : undefined;
    this.egyptianValidator = layers.has('EGYPTIAN') ? new EgyptianBundleValidator() : undefined;
  }

  /**
   * Run the pipeline against an inbound request. Throws the
   * appropriate typed `HfcxError` on validation failure; resolves with
   * a {@link RecipientResult} on success.
   */
  async handle(
    authorizationHeader: string | null | undefined,
    protocolHeaders: Readonly<Record<string, string>>,
    requestBody: string,
  ): Promise<RecipientResult> {
    if (!protocolHeaders) throw new TypeError('protocolHeaders is required');
    if (typeof requestBody !== 'string') throw new TypeError('requestBody must be a string');

    const correlationId = protocolHeaders[PROTOCOL_HEADER_CORRELATION_ID] ?? 'no-correlation-id';

    return runWithCorrelationId(correlationId, async () => {
      if (this.enabledLayers.has('BEARER')) {
        this.bearerValidator!.validate(authorizationHeader);
      }
      if (this.enabledLayers.has('HEADERS')) {
        this.headerValidator!.validate(protocolHeaders);
      }

      const jwe = extractPayload(requestBody);
      const decrypted = await this.decryptor.decrypt(jwe);

      this.fhirValidator?.validate(decrypted);
      this.egyptianValidator?.validate(decrypted);

      return {
        decryptedPayload: decrypted,
        protocolHeaders,
        correlationId,
      };
    });
  }
}

function extractPayload(requestBody: string): string {
  let parsed: unknown;
  try {
    parsed = JSON.parse(requestBody);
  } catch (err) {
    throw new EnvelopeMalformedJsonError(
      `Request body is not valid JSON: ${(err as Error).message}`,
    );
  }
  if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new EnvelopeMissingPayloadError('Request body envelope is not a JSON object');
  }
  const payload = (parsed as { payload?: unknown }).payload;
  if (typeof payload !== 'string' || payload.length === 0) {
    throw new EnvelopeMissingPayloadError("Request body envelope missing required 'payload' field");
  }
  return payload;
}
